synapse/lib/oauth.py

```diff
@@ -0,0 +1,398 @@
+                    continue
+
+                ok, data = await self._refreshAccessToken(providerconf, clientconf)
+                if not ok:
+                    logger.warning(f'OAuth V2 token refresh failed provider={provideriden} user={useriden} data={data}')
+
+                await self._setClientTokenData(provideriden, useriden, data)
+                self._schedule_item_ran.set()
+
+            self._schedule_empty.set()
+            await s_coro.event_wait(self.schedule_wake)
+            self.schedule_wake.clear()
+
+    def _normTokenData(self, issued_at, data):
+        '''
+        Normalize timestamps to be in epoch millis and set expires_at/refresh_at.
+        '''
+        reqValidTokenResponse(data)
+        expires_in = data['expires_in']
+        return {
+            'access_token': data['access_token'],
+            'expires_in': expires_in,
+            'expires_at': issued_at + expires_in * 1000,
+            'refresh_at': issued_at + (expires_in * self.refresh_window) * 1000,
+            'refresh_token': data.get('refresh_token'),
+        }
+
+    async def _getAccessToken(self, providerconf, authcode, code_verifier=None):
+        formdata = aiohttp.FormData()
+        formdata.add_field('grant_type', 'authorization_code')
+        formdata.add_field('scope', providerconf['scope'])
+        formdata.add_field('redirect_uri', providerconf['redirect_uri'])
+        formdata.add_field('code', authcode)
+        if code_verifier is not None:
+            formdata.add_field('code_verifier', code_verifier)
+
+        auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
+        return await self._fetchToken(providerconf['token_uri'], auth, formdata)
+
+    async def _refreshAccessToken(self, providerconf, clientconf):
+        refresh_token = clientconf['refresh_token']
+
+        formdata = aiohttp.FormData()
+        formdata.add_field('grant_type', 'refresh_token')
+        formdata.add_field('refresh_token', refresh_token)
+
+        auth = aiohttp.BasicAuth(providerconf['client_id'], password=providerconf['client_secret'])
+        ok, data = await self._fetchToken(providerconf['token_uri'], auth, formdata, retries=3)
+        if ok and not data.get('refresh_token'):
+            # if a refresh_token is not provided in the response persist the existing token
+            data['refresh_token'] = refresh_token
+
+        return ok, data
+
+    async def _fetchToken(self, url, auth, formdata, retries=1):
+
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Accept': 'application/json',
+        }
+
+        attempts = 0
+        issued_at = s_common.now()
+
+        async with aiohttp.ClientSession(timeout=self.timeout) as sess:
+
+            while True:
+                attempts += 1
+
+                try:
+                    async with sess.post(url, auth=auth, headers=headers, data=formdata, ssl=self.ssl) as resp:
```

Comment:
I do not believe we can handle a oauth target which is using self-signed certificates. While unusual, it is a bit of a departure from some other URL target based apis where we allow someone to create a struct that disables ssl verification. I could see this being used for a internal / test service which may not have trusted ( either root certs or via tls:ca:dir ) certificates.

Reply:
Added ssl verification as an option